src/Ledger/Fees.lagda

```diff
@@ -0,0 +1,94 @@
+\begin{code}[hide]
+{-# OPTIONS --safe #-}
+
+open import Ledger.Prelude hiding (_%_; _*_)
+open import Ledger.Abstract
+open import Ledger.Transaction
+
+module Ledger.Fees
+  (txs : _) (open TransactionStructure txs)
+  (abs : AbstractFunctions txs) (open AbstractFunctions abs)
+  where
+
+open import Data.Rational using (0ℚ; ℚ; mkℚ+; _*_; floor)
+open import Data.Nat.Coprimality using (Coprime; 1-coprimeTo)
+open import Data.Nat.Properties using (<⇒<′; <′⇒<; ≰⇒>)
+-- open import Data.Nat.Properties.Ext using (suc∸≤′)
+open import Data.Integer using (∣_∣)
+open import Data.Product using (swap)
+\end{code}
+
+The function \AgdaFunction{scriptsCost} calculates the fee for reference scripts in the transaction
+using a function that is piece-wise linear in the size, where the linear constant multiple grows with each
+\refScriptCostStride bytes.  Thus, the \AgdaFunction{scriptsCost} function depends on the
+\AgdaFunction{scriptsTotalSize} function, which returns an integer that is the total size of
+the reference script in bytes, as well as the following protocol parameters:
+\begin{itemize}
+\item \refScriptCostMultiplier, a rational number, the growth factor or step multiplier that determines how
+much the price per byte increases after each increment;
+\item \refScriptCostStride, an integer, the size in bytes at which the price per byte grows linearly;
+\item \minFeeRefScriptCoinsPerByte, a rational number, the base fee or initial price per byte.
+\end{itemize}
+
+\begin{code}[hide]
+_↑ℚ : ℕ → ℚ
+n ↑ℚ = mkℚ+ n 1 (1-coprimeTo n ∘ swap)
+
+data Terminates (x : ℕ) : Set where
+  termination-proof : (∀ y → (y <′ x) → Terminates y) → Terminates x
+
+
+base-case : (∀ y → (y <′ zero) → Terminates y)
+base-case _ ()
+
+generateTerminationProof : (n : ℕ) → Terminates n
+
+induction : (n : ℕ) → (y : ℕ) → (y <′ n) → Terminates y
+induction n zero _ = termination-proof base-case
+induction .(suc (suc y')) (suc y') ≤′-refl =  generateTerminationProof (suc y')
+induction .(suc n) (suc y') (≤′-step {n} m≤′n) = induction n (suc y') m≤′n
+
+generateTerminationProof zero = termination-proof (base-case)
+generateTerminationProof (suc limit) = termination-proof (induction (suc limit))
```

Comment:
Note that there is `Data.Nat.Induction` for this stuff, no need to do it by hand